js_modules/dagster-ui/packages/ui-core/src/runs/HourlyDataCache.tsx

```diff
@@ -0,0 +1,205 @@
+    }
+
+    return missingIntervals;
+  }
+
+  /**
+   * Checks if a range is completely cached.
+   * @param start - The start time in milliseconds.
+   * @param end - The end time in milliseconds.
+   * @returns True if the range is completely cached, false otherwise.
+   */
+  isCompleteRange(start: number, end: number): boolean {
+    const startHour = Math.floor(start / ONE_HOUR);
+    const endHour = Math.floor(end / ONE_HOUR);
+
+    if (startHour !== endHour) {
+      throw new Error('Expected the input range to be within a single hour');
+    }
+
+    if (this.cache.has(startHour)) {
+      const intervals = this.cache.get(startHour)!;
+      let currentStart = start;
+
+      for (const {start: cachedStart, end: cachedEnd} of intervals) {
+        if (cachedStart > currentStart) {
+          return false;
+        }
+        if (cachedEnd >= end) {
+          return true;
+        }
+        currentStart = Math.max(currentStart, cachedEnd);
+      }
+
+      return currentStart >= end;
+    }
+
+    return false;
+  }
+
+  /**
+   * Merges overlapping intervals.
+   * @param intervals - The intervals to merge.
+   * @returns An array of merged intervals.
+   */
+  private mergeIntervals(intervals: Array<TimeWindow<T>>): Array<TimeWindow<T>> {
+    if (intervals.length === 0) {
+      return [];
+    }
+
+    intervals.sort((a, b) => a.start - b.start);
+    const mergedIntervals: Array<TimeWindow<T>> = [intervals[0]!];
+
+    for (const current of intervals.slice(1)) {
+      const lastMerged = mergedIntervals[mergedIntervals.length - 1]!;
+
+      if (current.start <= lastMerged.end) {
+        lastMerged.end = Math.max(lastMerged.end, current.end);
+        lastMerged.data = lastMerged.data.concat(current.data);
+      } else {
+        mergedIntervals.push(current);
+      }
+    }
+
+    return mergedIntervals;
+  }
+}
+
+/**
+ * Breaks a time range into hourly buckets.
+ * @param startTime - The start time in milliseconds.
+ * @param endTime - The end time in milliseconds.
+ * @returns An array of [start, end] pairs representing each hourly bucket.
+ */
+export function getHourlyBuckets(startTime: number, endTime: number): [number, number][] {
+  const buckets: [number, number][] = [];
+
+  // Convert start and end times to the number of hours since epoch
+  const startHour = Math.floor(startTime / ONE_HOUR) * ONE_HOUR;
+
+  // Handle the first partial bucket
+  if (startTime !== startHour) {
+    const firstBucketEnd = startHour + ONE_HOUR - 1;
+    buckets.push([startTime, Math.min(firstBucketEnd, endTime)]);
+  }
+
+  // Add full hourly buckets
+  let currentStart = startHour + (startTime === startHour ? 0 : ONE_HOUR);
+  while (currentStart + ONE_HOUR <= endTime) {
+    const nextHour = currentStart + ONE_HOUR;
+    buckets.push([currentStart, nextHour - 1]);
+    currentStart = nextHour;
+  }
+
+  // Handle the last partial bucket
+  if (currentStart < endTime) {
+    buckets.push([currentStart, endTime]);
+  }
+
```

Comment:
One thing that would simplify it a bit is always fetching full hours -- I don't know that it's worth fetching "44 minutes, hour, hour, hour, 16 minutes" vs. "hour, hour, hour, hour (partially in the future)"

Doing full hour fetches all the time would mean not having to think about `isCompleteRange` or `getMissingIntervals` and make this a fair bit simpler I think.

If `always-full-hours` causes too much overfetching, you could make the windows smaller eg `always-full-15min`.

Either way the code looks clean and with tests this should be easy to maintain!